source/slang/slang-check-conversion.cpp

```diff
@@ -203,6 +203,292 @@ DeclRef<StructDecl> findBaseStructDeclRef(
     return baseStructDeclRef;
 }
 
+ConstructorDecl* SemanticsVisitor::_getSynthesizedConstructor(
+    StructDecl* structDecl,
+    ConstructorDecl::ConstructorTags tags)
+{
+    ConstructorDecl* synthesizedCtor = nullptr;
+    if (structDecl->m_synthesizedCtorMap.tryGetValue((int)tags, synthesizedCtor))
+    {
+        return synthesizedCtor;
+    }
+
+    return nullptr;
+}
+
+static StructDecl* _getStructDecl(Type* type)
+{
+    if (as<VectorExpressionType>(type) || as<MatrixExpressionType>(type) ||
+        as<ArithmeticExpressionType>(type) || as<BuiltinType>(type))
+    {
+        return nullptr;
+    }
+
+    if (auto declRefType = as<DeclRefType>(type))
+    {
+        auto structDecl = as<StructDecl>(declRefType->getDeclRef());
+        return structDecl.getDecl();
+    }
+
+    return nullptr;
+}
+
+bool SemanticsVisitor::_cStyleStructBasicCheck(Decl* decl)
+{
+    // 1. It has to be a user-defined struct type, or a basic scalar, vector or matrix type
+    StructDecl* structDecl = nullptr;
+    if (isFromCoreModule(decl))
+        return false;
+
+    if (auto varDecl = as<VarDecl>(decl))
+    {
+        auto type = varDecl->getType();
+        if (as<VectorExpressionType>(type) || as<MatrixExpressionType>(type) ||
+            as<BasicExpressionType>(type) || isDeclRefTypeOf<EnumDecl>(type).getDecl())
+            return true;
+
+        // check for user-defined struct type
+        structDecl = _getStructDecl(type);
+        if (!structDecl)
+            return false;
+    }
+
+    if (!structDecl)
+        structDecl = as<StructDecl>(decl);
+
+    // 2. It cannot have inheritance, but inherit from interface is fine.
+    for (auto inheritanceDecl : structDecl->getMembersOfType<InheritanceDecl>())
+    {
+        if (!isDeclRefTypeOf<InterfaceDecl>(inheritanceDecl->base.type))
+        {
+            return false;
+        }
+    }
+
+    // 3. It cannot have explicit constructor
+    if (_hasExplicitConstructor(structDecl, true))
+        return false;
+
+    // 4. All of its members have to have the same visibility as the struct itself.
+    DeclVisibility structVisibility = getDeclVisibility(structDecl);
+    for (auto varDecl : structDecl->getMembersOfType<VarDeclBase>())
+    {
+        if (getDeclVisibility(varDecl) != structVisibility)
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool SemanticsVisitor::isCStyleStruct(StructDecl* structDecl)
```

Comment:
It's weird to have the array case separated out from the structs.

I think the code should be restructured to recurse on type instead of decls. Basically you just need:
```
bool isCStyleType(Type* type);
```
An array can be a C-style type if it is sized and the element type is cstyle. A struct can be C-style if all the 5 conditions are met, etc.

This way we can easily extend the logic to cover more types than just structs as c-style if we want, and we can easily return false if the type is scalar,vector,matrix or enum types.